src/aspire/abinitio/commonline_sync3n.py

```diff
@@ -0,0 +1,1017 @@
+        :param b: distribution parameter
+        :param x0: Initial guess
+        :return: (log indicative probabilities, log arbitrary probabilities)
+        """
+        # These param values are passed to C, force doubles.
+        params = np.array([P2, A, a, B, b, x0], dtype=np.float64)
+
+        # host/gpu dispatch
+        if self.__gpu_module:
+            ln_f_ind, ln_f_arb = self._pairs_probabilities_cupy(Rijs, *params)
+        else:
+            ln_f_ind, ln_f_arb = self._pairs_probabilities_host(Rijs, *params)
+
+        return ln_f_ind, ln_f_arb
+
+    def _pairs_probabilities_host(self, Rijs, P2, A, a, B, b, x0):
+        """
+        See _pairs_probabilities.
+
+        CPU implementation.
+        """
+        # The following is adopted from Matlab pairs_probabilities_mex.c `looper`
+
+        # Initialize probability result arrays
+        ln_f_ind = np.zeros(len(Rijs), dtype=Rijs.dtype)
+        ln_f_arb = np.zeros(len(Rijs), dtype=Rijs.dtype)
+
+        c = np.empty((4), dtype=Rijs.dtype)
+        for i in trange(self.n_img - 2, desc="Computing pair probabilities"):
+            for j in range(i + 1, self.n_img - 1):
+                ij = self._pairs_to_linear[i, j]
+                Rij = Rijs[ij]
+                for k in range(j + 1, self.n_img):
+                    ik = self._pairs_to_linear[i, k]
+                    jk = self._pairs_to_linear[j, k]
+                    Rik = Rijs[ik]
+                    Rjk = Rijs[jk]
+
+                    # Compute conjugated rotats
+                    Rij_J = J_conjugate(Rij)
+                    Rik_J = J_conjugate(Rik)
+                    Rjk_J = J_conjugate(Rjk)
+
+                    # Compute R muls and norms
+                    c[0] = np.sum(((Rij @ Rjk) - Rik) ** 2)
+                    c[1] = np.sum(((Rij_J @ Rjk) - Rik) ** 2)
+                    c[2] = np.sum(((Rij @ Rjk_J) - Rik) ** 2)
+                    c[3] = np.sum(((Rij @ Rjk) - Rik_J) ** 2)
+
+                    # Find best match
+                    best_i = np.argmin(c)
+                    best_val = c[best_i]
+
+                    # For each triangle side, find the best alternative
+                    alt_ij_jk = c[self._ALTS[0][best_i][0]]
+                    if c[self._ALTS[1][best_i][0]] < alt_ij_jk:
+                        alt_ij_jk = c[self._ALTS[1][best_i][0]]
+                    alt_ik_jk = c[self._ALTS[0][best_i][1]]
+                    if c[self._ALTS[1][best_i][1]] < alt_ik_jk:
+                        alt_ik_jk = c[self._ALTS[1][best_i][1]]
+                    alt_ij_ik = c[self._ALTS[0][best_i][2]]
+                    if c[self._ALTS[1][best_i][2]] < alt_ij_ik:
+                        alt_ij_ik = c[self._ALTS[1][best_i][2]]
+
+                    # Compute scores
+                    s_ij_jk = 1 - np.sqrt(best_val / alt_ij_jk)
+                    s_ik_jk = 1 - np.sqrt(best_val / alt_ik_jk)
+                    s_ij_ik = 1 - np.sqrt(best_val / alt_ij_ik)
```

Comment:
Looks like up to this point, we have the same code as `_triangle_scores_inner_host`. Would it make sense to factor it out into its own function?

Reply:
They are very similar but how/which various vars are used is slightly different.  I did successfully factor it out, but it isn't clear to me whether it is cleaner or just more abstract.  

These loops are super time consuming on cpu... to avoid constantly initializing arrays I pass them by ref (which is potentially confusing) instead of functionally returning them.  It also is getting farther from following the MATLAB code, for whatever that is worth.

Its a single commit so I can revert it if needed. LMK.

Reply:
No that's fine. Let's just keep as is. I didn't realize there were some differences.

Reply:
Ok I'll revert that commit then merge.